Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/organizer/poster/QrCodePosterViewModel.kt

```diff
@@ -0,0 +1,119 @@
+package de.rki.coronawarnapp.ui.eventregistration.organizer.poster
+
+import android.graphics.Bitmap
+import android.graphics.pdf.PdfDocument
+import android.view.View
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedFactory
+import dagger.assisted.AssistedInject
+import de.rki.coronawarnapp.eventregistration.checkins.qrcode.PosterTemplateProvider
+import de.rki.coronawarnapp.eventregistration.checkins.qrcode.QrCodeGenerator
+import de.rki.coronawarnapp.eventregistration.checkins.qrcode.Template
+import de.rki.coronawarnapp.eventregistration.storage.repo.TraceLocationRepository
+import de.rki.coronawarnapp.exception.ExceptionCategory
+import de.rki.coronawarnapp.exception.reporting.report
+import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
+import de.rki.coronawarnapp.util.files.FileSharing
+import de.rki.coronawarnapp.util.ui.SingleLiveEvent
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModel
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModelFactory
+import timber.log.Timber
+import java.io.File
+import java.io.FileOutputStream
+import java.lang.ref.WeakReference
+
+class QrCodePosterViewModel @AssistedInject constructor(
+    @Assisted private val traceLocationId: Long,
+    private val dispatcher: DispatcherProvider,
+    private val qrCodeGenerator: QrCodeGenerator,
+    private val posterTemplateProvider: PosterTemplateProvider,
+    private val traceLocationRepository: TraceLocationRepository,
+    private val fileSharing: FileSharing
+) : CWAViewModel(dispatcher) {
+
+    private val posterLiveData = MutableLiveData<Poster>()
+    val poster: LiveData<Poster> = posterLiveData
+    val sharingIntent = SingleLiveEvent<FileSharing.FileIntentProvider>()
+
+    init {
+        generatePoster()
+    }
+
+    /**
+     * Create a new PDF file and result is delivered by [sharingIntent]
+     * as a sharing [FileSharing.ShareIntentProvider]
+     */
+    @Suppress("BlockingMethodInNonBlockingContext")
+    fun createPDF(view: View) = launch(context = dispatcher.IO) {
+        try {
+            val weakViewRef = WeakReference(view) // Accessing view in background thread
+            val directory = File(view.context.cacheDir, "poster").apply { if (!exists()) mkdirs() }
+            val file = File(directory, "cwa-qr-code.pdf")
+
+            val weakView = weakViewRef.get() ?: return@launch // View is not existing anymore
```

Comment:
Do you think the view could not exist anymore because the user has left the screen? 

What about calling `ensureActive()` instead to check if the viewModelScope wasn't cancelled already?

Reply:
It is not about the ViewModel ,the view model survives in configuration changes. So it could be that the fragment view is destroyed ,but not the ViewModel. this one point, the other point is the some work is being done in another thread and user could decide to leave the screen and the view will destroyed too. to avoid such cases I hold a weak reference to the view just in case